src/lib/coda_base/pending_coinbase.ml

```diff
@@ -570,41 +719,111 @@ struct
       in
-      handle
-        (Merkle_tree.modify_req ~depth (Hash.var_to_hash_packed t) addr
-           ~f:(fun stack0 ->
-             let total_coinbase_amount =
-               Currency.Amount.var_of_t Coda_compile_config.coinbase
-             in
-             let%bind rem_amount =
-               Currency.Amount.Checked.sub total_coinbase_amount amount
-             in
-             let%bind amount1_equal_to_zero = equal_to_zero amount in
-             let%bind amount2_equal_to_zero = equal_to_zero rem_amount in
-             (* TODO: Optimize here since we are pushing twice to the same stack *)
-             let%bind stack_with_amount1 =
-               Stack.Checked.push stack0 (pk, amount, state_body_hash)
-             in
-             let%bind stack_with_amount2 =
-               Stack.Checked.push stack_with_amount1
-                 (pk, rem_amount, state_body_hash)
-             in
-             chain Stack.if_ amount1_equal_to_zero ~then_:(return stack0)
-               ~else_:
-                 (Stack.if_ amount2_equal_to_zero ~then_:stack_with_amount1
-                    ~else_:stack_with_amount2) ))
-        reraise_merkle_requests
-      >>| Hash.var_of_hash_packed
+      let%bind no_update = Update.Action.Checked.no_update action in
+      let update_state_stack (stack : Stack.var) =
+        (*get previous stack to carry-forward the stack of state body hashes*)
+        let%bind previous_state_stack =
+          request_witness Coinbase_stack_state.typ
+            As_prover.(map (return ()) ~f:(fun () -> Get_previous_stack))
+        in
+        let stack_initialized = {stack with state= previous_state_stack} in
+        let%bind stack_with_state_hash =
+          Stack.Checked.push_state state_body_hash stack_initialized
+        in
+        (*Always update the state body hash unless there are no transactions in this block*)
+        Stack.Checked.if_ no_update ~then_:stack ~else_:stack_with_state_hash
+      in
+      let f1 stack0 =
+        let%bind stack = update_state_stack stack0 in
+        let total_coinbase_amount =
+          Currency.Amount.var_of_t Coda_compile_config.coinbase
+        in
+        let%bind rem_amount =
+          Currency.Amount.Checked.sub total_coinbase_amount amount
+        in
+        let%bind no_coinbase_in_this_stack =
+          Update.Action.Checked.update_two_stacks_coinbase_in_second action
+        in
+        let%bind amount1_equal_to_zero = equal_to_zero amount in
+        let%bind amount2_equal_to_zero = equal_to_zero rem_amount in
+        let%bind () =
+          with_label __LOC__
+            (let%bind check1 = Boolean.equal no_update amount1_equal_to_zero in
+             Boolean.Assert.all [check1])
+        in
+        let%bind no_coinbase =
+          let%bind no_update = Boolean.(no_update && amount1_equal_to_zero) in
+          Boolean.(no_update || no_coinbase_in_this_stack)
+        in
+        (* TODO: Optimize here since we are pushing twice to the same stack *)
```

Comment:
Is this TODO still relevant?

Reply:
Yea, this is because we push two coinbase with the same proposer pk to the same stack